### PR TITLE
feat(NavBarLink)!: renamed to NavBarButton; improved styles

### PR DESCRIPTION
**Pull request by slingner**
Related to issue #74

- Updated `NavbarLink` to now be `NavBarButton`.
- Made a new story specifically for NavBarButton (took out of navbar story), added variations for all States.

### CODE REVIEW

File: libs/spark/src/NavBarItem.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarItem.tsx:13]: `interface` not `type`, extend Mui's `ButtonProps`, omit `variant`, `color`. see CONTRIBUTING.md for a template on this

File: libs/spark/src/NavBarItem.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarItem.tsx:46]: Not sure if you want this to be commented on, but since this component doesn't have >1 variant, we can avoid going along the same lines as Spark's current `Button` implementation. You can use Mui's `Button` directly. Note that you're already doing this based on what you're styling above.

File: libs/spark/src/NavBarItem.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarItem.tsx:21]: You don't need to redeclare ones you want. You also don't need to define href, a consumer can do

```
<NavBarItem component="a" href="..." />
```
and that will fine by Mui's default implementation. Likewise for react-router
```
<NavBarItem component={RouterLink} to="..." />
```

See https://material-ui.com/components/buttons/#third-party-routing-library

Comment (slingner) [reply libs/spark/src/NavBarItem.tsx:21]: Thanks, William, that's good to know about the React Router implementation. I will need to put some custom props in this interface though, correct? Maybe just children?

Comment (slingner) [reply libs/spark/src/NavBarItem.tsx:21]: Without any, I'm getting an empty interface warning.

Comment (slingner) [reply libs/spark/src/NavBarItem.tsx]: Two Issues:
1. I'm still getting an error message when trying to call `component={Link}`
2. Also trying to work out how to target the clicked state of the button. I attempted using the `aria-current='true'` in styles like Joshua had previously, but was not successful as of yet.

Comment (WilliamKelley) [reply libs/spark/src/NavBarItem.tsx]: I wouldn't add react-router and demonstrate the integration. It's more likely to cause problems than anything. As long as we're extending Button, then we can absorb Mui's test suite and be assured that the integration will still work. I'd recommend just showing `<NavBarItem component="a" href="..." />` use case.

You won't need to add children, again that's already defined. Yeah I'm struggling how to set it equal to the other interface. Maybe ask around

File: libs/spark/src/NavBarItem.tsx
Comment (slingner) [reply libs/spark/src/NavBarItem.tsx]: What's the harm of doing a type here? I'm not getting an error when using a type when using Omit

```
export type NavBarItemProps = Omit<ButtonProps, 'variant' | 'color' | 'size'>;
// export interface NavBarItemProps
//   extends Omit<ButtonProps, 'variant' | 'color' | 'size'> {}
```

File: libs/spark/src/NavBarItem.tsx
Comment (slingner) [reply libs/spark/src/NavBarItem.tsx]: What's the harm of doing a type here? I'm not getting an error when using a type when using Omit

```
export type NavBarItemProps = Omit<ButtonProps, 'variant' | 'color' | 'size'>;
// export interface NavBarItemProps
//   extends Omit<ButtonProps, 'variant' | 'color' | 'size'> {}
```

Comment (WilliamKelley) [reply libs/spark/src/NavBarItem.tsx]: I'm not sure. One consequence is that a consumer can extend a type, but they can extend an interface. I know they can union a type and all that jazz, but I feel like there must be a way to do it with interfaces. Maybe just setting the component type to be `ButtonProps`?

Comment (slingner) [reply libs/spark/src/NavBarItem.tsx]: I haven't solved this piece but I have made a lot of progress on matching the Figma design today. Curious to hear your thoughts on its current state. I'll keep researching on extending interfaces in the meantime.

File: libs/spark/stories/navbar.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbar.stories.tsx:33]: a

File: libs/spark/stories/navbar.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbar.stories.tsx:32]:
```suggestion
      <NavBarItem href="/" startIcon={<HomeIconDuotone fillColor="transparent" />}>
        Dashboard
      </NavBarItem>
```

Comment (WilliamKelley) [reply libs/spark/stories/navbar.stories.tsx:32]: likewise for all the rest

File: libs/spark/stories/navbar.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbar.stories.tsx:61]: This should be in it's own story

File: libs/spark/stories/navbar.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbar.stories.tsx:55]: Once in its own story, simplify to `Template`

File: libs/spark/stories/navbar.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbar.stories.tsx:61]: Once in it's own story, simplify to `Basic`

File: libs/spark/test/navbar.spec.tsx
Comment (WilliamKelley) [inline libs/spark/test/navbar.spec.tsx:11]: same change as in the story

**Review by WilliamKelley — CHANGES_REQUESTED** [review]
Looking good! i'll go through the styling more carefully after this feedback is addressed.

**Review by slingner — COMMENTED** [review]
> Looking good! i'll go through the styling more carefully after this feedback is addressed.

Updated! So much cleaner putting the `startIcon` directly in and deleting the span wrapper.

**Review by WilliamKelley — CHANGES_REQUESTED** [review]
I thought about the name this weekend, and I think we should go with `NavBarButton`. Reason being, it's already within Mui's Button API to be an anchor tag if an href is provided. And the only other ideal that I had in mind was that an `Item` could be a parent container of whatever component made sense to be the child. But we can get what we want just from the Button.

File: libs/spark/stories/navbar.stories.tsx
Comment (slingner) [reply libs/spark/stories/navbar.stories.tsx:31]: > Carry to rest of icons

Having trouble with this one, do we need to do a theme override? 

[https://stackoverflow.com/questions/62677833/change-size-of-starticon-endicon-material-ui]

Comment (WilliamKelley) [reply libs/spark/stories/navbar.stories.tsx]: No need for theme overrides, anything we can do there we can do here. I'll note though, I think my theme overrides to the SvgIcon might have been heavy handed and making life more difficult than necessary here in places -- but that's for another time, and not affecting this specific comment directly.

See the stylesheet below. There's a style rule with higher specificity than the rule that gets applied when `fontSize="large"` (i.e. the one setting `font-size: 2rem`). So you just have to match that winning rule to have your change take effect. We want to do that -- fixing intended prop behavior -- rather than being heavy handed in that rule by overriding every instance of an icon to be `2rem`.
image
So something like
```

  iconSizeMedium: {
    '& > *:first-child': {
      fontSize: '1.5rem', // FYI, our small/medium/large for icons is 16px/24px/32px
      '&.MuiSvgIcon-fontSizeLarge': {
        fontSize: '2rem',
      },
    },
  },
```

File: libs/spark/stories/navbar.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbar.stories.tsx:30]:
```suggestion
      <NavBarButton href="#" startIcon={<HomeIconDuotone />} aria-current="page">
```
`href`'s shouldn't be blank. To not go anywhere, just supply a hash.

**Review by WilliamKelley — CHANGES_REQUESTED** [review]
We can better validate styles if you expand on your stories for `NavBarButton`. 

Follow the pattern of some other components with expanded stories and create ones

- Patterns (label only, start icon plus label, label plus end icon)
- Patterns Disabled
- Patterns Hover
- Patterns Focus
- Patterns Active

Patterns is arbitrary, idk a better name to describe those three options

**Comment by slingner** [comment]
> We can better validate styles if you expand on your stories for `NavBarButton`.
> 
> Follow the pattern of some other components with expanded stories and create ones
> 
> * Patterns (label only, start icon plus label, label plus end icon)
> * Patterns Disabled
> * Patterns Hover
> * Patterns Focus
> * Patterns Active
> 
> Patterns is arbitrary, idk a better name to describe those three options

I'm not fully understanding how to target the hover and focused states here? I see them being implemented in the story, but it doesn't appear to be showing those states unless you actually do the hover?  Maybe I'm missing something here.

**Comment by WilliamKelley** [comment]
> I'm not fully understanding how to target the hover and focus states here? I see them being implemented in the story, but it doesn't appear to be showing those states unless you actually do the hover/focus? Maybe I'm missing something here.

You probably need to `npm i`

File: libs/spark/stories/navbarbutton.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbarbutton.stories.tsx:29]: Group in a single template and make sure to spread `args` across each. then
```tsx
export const All = Template.bind({});
export const AllDisabled = Template.bind({});
AllDisabled.args = { disabled: true };
export const AllHover = Template.bind({});
AllHover.parameters = { pseudo: { hover: true } };
export const AllFoucs = Template.bind({});
AllFocus.parameters = { pseudo: { focus: true } };
```

File: libs/spark/stories/navbarbutton.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbarbutton.stories.tsx:60]: Leave these labels off, they're not useful in the scenario

File: libs/spark/stories/navbarbutton.stories.tsx
Comment (WilliamKelley) [inline libs/spark/stories/navbarbutton.stories.tsx:61]: These need to be wrapped in <span>'s because the grid will stretch its children to take up their entire row/column space, but we don't want that.

**Review by WilliamKelley — CHANGES_REQUESTED** [review]
Aside from the grid stretching issue, the stylings don't seem to match figma exactly, ex: disabled icon

**Comment by slingner** [comment]
> Aside from the grid stretching issue, the stylings don't seem to match figma exactly, ex: disabled icon

Good catch on this one, just updated.

**Review by WilliamKelley — COMMENTED** [review]
I still see styling issues with inconsistent padding, especially around the icons. They seem to have weird margins be default. For this, I'd wait until #144 is merged, then revisit. should be today.

Change your PR title from item to button.

The focus border is behaving poorly. 
    - First, since there's no border on the button by default, the appearance of a focus border causes layout shifts in surrounding content. I've fixed this in other place by setting a border on the default style, but the same color as whatever the background is supposed to be in each state/pseudo-state. Then when it needs to be a different color, like in focus, then just change `borderColor`. No layout shifts. Also requires accounting for that 2px width in things like padding.
    - Second, it seems to have a default transition from Mui on it that's just laggy. Maybe it's because of the layout shifts above and fixing that will fix this, but if not, just declare the transitions (copy paste from Mui), but without targeting border.

Lastly, it seems like if the only prescribed size for NavBarButton is 48px in height, then it should take on the default prop of `size="large"` since that's the same height. Some of your stylings will have to target different class slots after changing that, but you shouldn't have to declare a specific height as a result. Or ideally even fontsize or anything.

File: libs/spark/src/NavBarButton.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarButton.tsx:6]: Unless we can figure out a way to associate this with the declaration of `NavBarButton` then its useless as anything except a reference.

File: libs/spark/src/NavBarButton.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarButton.tsx:13]: since border is 2px, padding has to account for that
```suggestion
    padding: 6,
```

File: libs/spark/src/NavBarButton.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarButton.tsx:21]: it looks like this isn't necessary after all. I tried commenting it out and it no longer lags, so it was due to the progressive layout shifts solved by a constant border size.
```suggestion
```

File: libs/spark/src/NavBarButton.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarButton.tsx:43]:
```suggestion
        '& > *:first-child': {
```
Some icons have `<circle>` or other svg child tags.

File: libs/spark/src/NavBarButton.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarButton.tsx:67]:
```suggestion
        '& > *:first-child': {
```

File: libs/spark/src/NavBarButton.tsx
Comment (WilliamKelley) [inline libs/spark/src/NavBarButton.tsx:64]:
```suggestion
      borderColor: 'transparent',
```
Otherwise all sub-properties of `border` are defaulted to browser defaults. You could notice the button's size reduced when switch stories from All to All Disabled.

**Review by WilliamKelley — CHANGES_REQUESTED** [review]
See my comments.

In addition, we'll have to wait on this PR until some issues with Button are fixed. The icons are still off:
current:
image
intended:
image

But that requires another PR.

Also, we've run out of Chromatic snapshots so we can't verify everything else is staying the same, so we gotta discuss that as a team.

**Comment by WilliamKelley** [comment]
Also, add another story for the `active` pseudo state

**Comment by WilliamKelley** [comment]
Keeping this as Draft until we verify locally that all styles match Figma, and this PR is ready for final chromatic review

**Comment by WilliamKelley** [comment]
@slingner Checking in on the status of this.

**Comment by slingner** [comment]
> @slingner Checking in on the status of this.

Thanks for checking in on this William, I had to step away to work on some Create UI updates, but jumping back in now. Just pushed up changes with the active state in the NavBarButton story. 

Should I do a rebase on main also?

**Comment by WilliamKelley** [comment]
@slingner Yes definitely should rebase. I'll take a look later today after that.

**Comment by slingner** [comment]
Superseding by PR #166

**Pull request closed without merging**
